### PR TITLE
ecshreve: [bug] - fix schema field definition for class level

### PR DESCRIPTION
**Pull request by ecshreve**
## What does this do?
The creating_spell_slots field is an array of subdocuments, and
in mongoose arrays implicitly have a default value of []. This was
causing this field to be included in the response for every resource
from this endpoint regardless of class.

Explicitly setting default: undefined fixes this.

see the "Arrays" section under "SchemaType Options": https://mongoosejs.com/docs/schematypes.html#schematype-options

## How was it tested?
localhost

## Is there a Github issue this is resolving?
no

## Was any impacted documentation updated to reflect this change?
fixing the documentation for this endpoint was how i found this bug :)

## Here's a fun image for your troubles
random photo - update me

### CODE REVIEW

File: src/models/level/index.js [hunk 1]
```diff
@@ -28,7 +28,10 @@ const ClassSpecific = new Schema({
   bardic_inspiration_die: { type: Number, index: true },
   brutal_critical_dice: { type: Number, index: true },
   channel_divinity_charges: { type: Number, index: true },
-  creating_spell_slots: [ClassSpecificCreatingSpellSlot],
+  creating_spell_slots: {
+    type: [ClassSpecificCreatingSpellSlot],
+    default: undefined,
```
Comment (bagelbits) [inline src/models/level/index.js:33]: Holy shit. I didn't know you could do this. This is actually a bug in all the other places we use arrays in nested objects.

Comment (bagelbits) [reply src/models/level/index.js:33]: Do you want to do the others while you're in here?

Comment (ecshreve) [reply src/models/level/index.js:33]: sure! i'll tidy the rest of them up this week sometime. we can either merge this pr now or leave it open and i can tack on fixes for any other instances i find. either way is fine by me, it's a super low impact bug so not too worried about it being there for a bit longer.

Comment (bagelbits) [reply src/models/level/index.js:33]: I'll just leave it open for now.

**Comment by ecshreve** [comment]
I did a little more thinking and tinkering with this, had a few thoughts...

tldr: explicitly going through each schema and overriding the default for any arrays to undefined would by and large have no effect on the shape of data returned from the api.

<hr>

_kept some notes while I was reading / poking around, this is my current understanding / thoughts, quite possible I'm misunderstanding some pieces, but figured I'd drop it here anyway._

By and large the structure of a document in a particular collection in the db, matches a mongoose schema defined under `src/models`. Schema field types can be grouped into 2 main buckets, each handles default values and undefined fields slightly differently. https://mongoosejs.com/docs/guide.html

**single value, either primitive or object** 
If the named field exists in the schema, but not on a found document, or the field explicitly has a value of `undefined` in the document, then mongoose returns the found document without that field.

**array of primitive or schema**
When a field exists in the schema, but not on a particular document then mongoose includes the field with a default value `[]` in the returned document. It's possible to override mongoose's defaulting behavior by setting a default in the schema definition, if the default is set to `undefined` then the field behaves like a single value field, and isn't included in the response if it doesn't exist on the found document.

For the most part this works out fine because when we're querying for documents from a particular collection we don't get any fields back that aren't expected, and all fields in the returned document and nested subdocuments make sense either way (not appearing in the response, or appearing with a value of `[]`). Additionally, in most cases it looks like we explicitly set array field values to `[]` at the db level (most of the time).  And most of the time when we have arrays or objects in nested fields we're treating it more like an inhertancy-ish thing, where any of the subsequent nested fields, while possiblly undefined or empty, is still related or applicable to an item of that schema.

The situation addressed in this PR is somewhat different though, in this case we are using a shared schema definition for an object we expect to have different shapes for different documents (class-specific attributes for a particular level). We don't explicitly enumerate defaults at the db level in this subdocument, which works because mongoose handles filtering out undefined fields based on the schema ... _BUT_  the `ClassSpecific` definition includes a field with an array of objects `creating_spell_slots:  [ClassSpecificCreatingSpellSlot]`which mongoose was setting to the default `[]` after retrieving the document, before returning the found data.

For this use case (polymorphic-esque use of a shared type definition) I think it definitely makes sense to explicitly set the default to `undefined` in the schema to prevent including unexpected fields in the response. But for most other situations it shouldn't be needed.

I think with our steady march toward typescriptifying things it might make sense to decide how we want to handle missing / undefined fields like this and find a way to do that in TS world. Also if / when we loop some e2e tests we'll have a better idea of if any other cases like this one currently exist. For example, I found this bug related to `ClassSpecific` with that postman thingy i was messing around with, because in the response from `/api/classes/barbarian/{level}`, under the `class_specific` field, in addition to the expected fields for a barbarian, there was an unexpected `creating_spell_slots: []` , and when validating that response against the `ClassLevel` swagger model it fails to satisfy the `AnyOf` condition on the field. I ran the postman tests for every resource that is accessible from an api path like this `/api/{resource}/{index}` (~2k items) and all of those passed the schema validation. with a little more tinkering should be able to validate all the resources from the other paths.

**Comment by bagelbits** [comment]
@ecshreve I specifically know that choice stuff will fall into this trap sometimes because it's nested. As well as a few other places. However, you are correct that TSifying stuff and e2e tests will eventually knock some of this loose. Also, you have the right of it.

**Review by bagelbits — APPROVED** [review]
